lib/rucio/core/transfer.py

```diff
@@ -702,6 +781,271 @@ def __rewrite_dest_url(dest_url, dest_sign_url, dest_scheme):
+            transfer['sources'] = [(source_rse_name, source_url, source_rse_id, source.source_ranking, computed_distance)]
+            transfer['account'] = rws.account
+            if rws.previous_attempt_id:
+                file_metadata['previous_attempt_id'] = rws.previous_attempt_id
+
+        transfer_path.append(transfer)
+    return transfer_path
+
+
+def __merge_or_discard_tranfer_definitions(candidate_paths):
+    if not candidate_paths:
+        return
+
+    # Discard multihop transfers which contain a tape source as an intermediate hop
+    filtered_candidate_paths = []
+    for path in candidate_paths:
+        if any(transfer['file_metadata']['src_type'] == 'TAPE' for transfer in path[1:]):
+            continue
+        filtered_candidate_paths.append(path)
+    candidate_paths = filtered_candidate_paths
+
+    # Discard multihop transfers which contain other candidate as part of itself For example:
+    # if A->B->C and B->C are both candidates, discard A->B->C because it includes B->C. Doing B->C is enough.
+    source_rses = {path[0]['file_metadata']['src_rse_id'] for path in candidate_paths}
+    filtered_candidate_paths = []
+    for path in candidate_paths:
+        if any(transfer['file_metadata']['src_rse_id'] in source_rses for transfer in path[1:]):
+            continue
+        filtered_candidate_paths.append(path)
+    candidate_paths = filtered_candidate_paths
+
+    def __transfer_order_key(transfer_path):
+        # higher source_ranking first,
+        # on equal source_ranking, prefer DISK over TAPE
+        # on equal type, prefer lower distance_ranking
+        # on equal distance, prefer single hop
+        return (
+            - transfer_path[-1]['sources'][0][3],
+            transfer_path[0]['file_metadata']['src_type'].lower(),  # rely on the fact that "disk" < "tape" in string order
+            transfer_path[-1]['sources'][0][4],
+            len(transfer_path) > 1,  # rely on the fact that False < True
+        )
+
+    best_candidate = min(candidate_paths, key=__transfer_order_key)
+
+    if best_candidate[0]['file_metadata']['src_type'] == 'TAPE':
+        # multiple Tape source replicas are not allowed in FTS3.
+        # Don't add any additional sources
+        return best_candidate
+
+    if len(best_candidate) > 1:
+        # only one source is allowed for multihop transfers to avoid spawning
+        # intermediate transfers all over the place
+        return best_candidate
+
+    for candidate in candidate_paths:
+        if candidate is best_candidate:
+            continue
+
+        if best_candidate[0]['file_metadata']['src_type'] != candidate[0]['file_metadata']['src_type']:
+            # Tape and Disk sources must not be used at the same time.
+            continue
+
+        if len(candidate) > 1:
+            # do not add multihop sources
+            continue
+
+        if not set(best_candidate[0]['schemes']).intersection(candidate[0]['schemes']):
```

Comment:
I have to think a bit more about this, but I am not 100% sure if this comparison is correct.
Because the `schemes` is already the compatible schemes of a specific chosen destination_scheme. But here you are iterating candidates (src-dest pairs) whose compatible_schemes is based on a different selected destination_scheme. I am not 100% sure if this is correct.